### PR TITLE
Add profiling options to submission_runner.py

### PR DESCRIPTION
**Pull request by pomonam**
This PR adds an option for profiling `submission_runner.py`. After this PR, you can add a flag `profile`, i.e,
```
python3 submission_runner.py \
    --framework=jax \
    --workload=mnist \
    --submission_path=baselines/mnist/mnist_jax/submission.py \
    --tuning_search_space=baselines/mnist/tuning_search_space.json \
    --profile
```
And the log will output some information on how much it took to sample a batch, update parameters, initialize the model, etc. Example output (with the command above) is shown below.
```
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Action                	|  Mean Duration (s)	|  Std Duration (s)	|  Num Calls      	|  Total Time (s) 	|  Percentage %   	|
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Total                 	|  -----          	|  -----          	|  5112           	|  14.15          	|  100 %          	|
----------------------------------------------------------------------------------------------------------------------------------------------------------
|  Train                 	|  13.083         	|  0.0            	|  1              	|  13.083         	|  92.46          	|
|  Data selection        	|  0.0022395      	|  0.0047329      	|  2553           	|  5.7174         	|  40.406         	|
|  Update parameters     	|  0.001238       	|  0.023317       	|  2553           	|  3.1606         	|  22.337         	|
|  Evaluation            	|  0.82529        	|  0.65484        	|  2              	|  1.6506         	|  11.665         	|
|  Initializing model    	|  1.0317         	|  0.0            	|  1              	|  1.0317         	|  7.2912         	|
|  Initializing dataset  	|  0.21838        	|  0.0            	|  1              	|  0.21838        	|  1.5433         	|
|  Initializing optimizer	|  0.061524       	|  0.0            	|  1              	|  0.061524       	|  0.43481        	|
----------------------------------------------------------------------------------------------------------------------------------------------------------
```

### CODE REVIEW

**Review by runame — APPROVED** [review]
Great work, this is really useful!

File: algorithmic_efficiency/profiler.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
+"""Profiling code for Jax and PyTorch
+
+Modified from:
+https://github.com/Lightning-AI/lightning/tree/master/src/pytorch_lightning/profiler
```
[the remark this answers is not in the document]
Comment (pomonam) [reply algorithmic_efficiency/profiler.py]: Thank you! Yes, you are right, fixed the reference.

File: submission_runner.py [hunk 2]
```diff
@@ -297,15 +313,23 @@ def score_submission_on_workload(workload: spec.Workload,
     rng = prng.PRNGKey(rng_seed)
     # If the submission is responsible for tuning itself, we only need to run it
     # once and return the total time.
-    score, _ = train_once(workload, global_batch_size, init_optimizer_state,
-                          update_params, data_selection, None, rng)
+    with profiler.profile('train'):
```
[the remark this answers is not in the document]
Comment (pomonam) [reply submission_runner.py]: Fixed!

**Review by znado — APPROVED** [review]
Looks great!

**Pull request merged**